library/Zend/ServiceManager/ServiceManager.php

```diff
@@ -718,9 +723,16 @@ public function has($name, $checkAbstractFactories = true, $usePeeringServiceMan
         }
 
         if ($usePeeringServiceManagers) {
+
+            $caller = $this->serviceManagerCaller;
             foreach ($this->peeringServiceManagers as $peeringServiceManager) {
-                if ($peeringServiceManager->has($name)) {
-                    return true;
+                // ignore peering service manager if they are the same instance
+                if ($caller !== $peeringServiceManager) {
+                    $peeringServiceManager->serviceManagerCaller = $this;
+                    if ($peeringServiceManager->has($name)) {
+                        return true;
```

Comment:
Shouldn't `$peeringServiceManager->serviceManagerCaller` be reset to `null` before returning?